modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableRow.java

```diff
@@ -388,37 +387,44 @@ private void setTreeItem(TreeItem<T> value) {
      *                                                                         *
      **************************************************************************/
 
-    private int index = -1;
     private boolean isFirstRun = true;
 
-    private void updateItem() {
+    private void updateItem(int oldIndex) {
         TreeTableView<T> tv = getTreeTableView();
         if (tv == null) return;
 
         // Compute whether the index for this cell is for a real item
-        boolean valid = index >=0 && index < tv.getExpandedItemCount();
+        final int newIndex = getIndex();
+        boolean valid = newIndex >= 0 && newIndex < tv.getExpandedItemCount();
 
         final TreeItem<T> oldTreeItem = getTreeItem();
         final boolean isEmpty = isEmpty();
 
         // Cause the cell to update itself
-        if (valid) {
+        outer: if (valid) {
             // update the TreeCell state.
             // get the new treeItem that is about to go in to the TreeCell
-            final TreeItem<T> newTreeItem = tv.getTreeItem(index);
+            final TreeItem<T> newTreeItem = tv.getTreeItem(newIndex);
             final T newValue = newTreeItem == null ? null : newTreeItem.getValue();
+            final T oldValue = oldTreeItem == null ? null : oldTreeItem.getValue();
 
             // For the sake of RT-14279, it is important that the order of these
             // method calls is as shown below. If the order is switched, it is
             // likely that events will be fired where the item is null, even
             // though calling cell.getTreeItem().getValue() returns the value
             // as expected
 
-            // There used to be conditional code here to prevent updateItem from
-            // being called when the value didn't change, but that led us to
-            // issues such as RT-33108, where the value didn't change but the item
-            // we needed to be listening to did. Without calling updateItem we
-            // were breaking things, so once again the conditionals are gone.
+            // RT-35864 - if the index didn't change, then avoid calling updateItem
+            // unless the item has changed.
+            if (oldIndex == newIndex) {
+                if (!isItemChanged(oldValue, newValue)) {
+                    // RT-37054:  we break out of the if/else code here and
+                    // proceed with the code following this, so that we may
+                    // still update references, listeners, etc as required.
+                    break outer;
```

Comment:
I understand neither the comment nor this `break outer;`: there is no code that follows `outer: if`, so we might as well `return`, right?

Reply:
Yes! I did it this way to be consistent with all other cell implementations. I really think that this should be refactored, I can also create a ticket for this. My preference would be to be consistent now, but change it for all cell implementations in one go.
What is your opinion?

Reply:
my opinion: the code is not broken per se.  Since this block is edited, I would have changed it just in this file to make the logic clearer (in other places, like TreeTableCell:685 it is functional because we have more code in line 692 and on).
I don't think another ticket is necessary because, technically speaking, the code is not broken.
(just my opinion)

Reply:
yes, but other cells have the exact same pattern, although it is not needed. And technically speaking, this is not even needed in `TreeTableCell`, tbh I never even needed a labeled if ever. 😄 
But sure, can also change.